network/openssl_privatekey.py

```diff
@@ -0,0 +1,246 @@
+            crypto_type = crypto.TYPE_DSA
+
+        try:
+            self.privatekey.generate_key(crypto_type, self.size)
+            privatekey_file = open(self.path, 'w')
+            privatekey_file.write(crypto.dump_privatekey(crypto.FILETYPE_PEM, self.privatekey))
+            privatekey_file.close()
+        except TypeError:
+            e = get_exception()
+            raise PrivateKeyError(e)
+        except ValueError:
+            e = get_exception()
+            raise PrivateKeyError(e)
+        except IOError:
+            self.remove()
+            e = get_exception()
+            raise PrivateKeyError(e)
+
+    def remove(self):
+        """Remove the private key from the filesystem."""
+
+        try:
+            os.remove(self.path)
+        except OSError:
+            e = get_exception()
+            if e.errno != errno.ENOENT:
+                raise PrivateKeyError(e)
+            else:
+                self.changed = False
+
+    def dump(self):
+        """Serialize the object into a dictionnary."""
+
+        result = {
+            'size': self.size,
+            'type': self.type,
+            'filename': self.path,
+            'changed': self.changed,
+        }
+
+        return result
+        
+
+def main():
+
+    module = AnsibleModule(
+        argument_spec = dict(
+            state=dict(default='present', choices=['present', 'absent'], type='str'),
+            size=dict(default=4096, type='int'),
+            type=dict(default='RSA', choices=['RSA', 'DSA'], type='str'),
+            force=dict(default=False, type='bool'),
+            path=dict(required=True, type='path'),
+        ),
+        supports_check_mode = True,
+    )
+
+    if not pyopenssl_found:
+        module.fail_json(msg='the python pyOpenSSL module is required')
+
+    path = module.params['path']
+    base_dir = os.path.dirname(module.params['path'])
+
+    if not os.path.isdir(base_dir):
+        module.fail_json(name=base_dir, msg='The directory %s does not exist or the file is not a directory' % base_dir)
+
+    private_key = PrivateKey(module)
+    if private_key.state == 'present':
+
+        if module.check_mode:
+            result = private_key.dump()
+            if not os.path.exists(path) or module.params['force']:
+                result['changed'] = True
+                module.exit_json(**result)
+            elif os.path.exists(path):
+                result['changed'] = False
+                module.exit_json(**result)
+
+        try:
+            private_key.generate()
+        except PrivateKeyError:
+            e = get_exception()
+            module.fail_json(msg='%s: %s' % (path, os.strerror(e.errno)))
+    else:
+
+        if module.check_mode:
+            result = private_key.dump()
+            if os.path.exists(path):
+                result['changed'] = True
+                module.exit_json(**result)
+            else:
+                result['changed'] = False
+                module.exit_json(**result)
```

Comment:
This can be simplified too:
```python
if module.check_mode:
    result = private_key.dump()
    result['changed'] = os.path.exists(path)
    module.exit_json(**result)
```